mount/mount_unix_test.go

```diff
@@ -230,3 +233,47 @@ func TestRecursiveUnmountTooGreedy(t *testing.T) {
 		t.Fatal("expected dir-other to be mounted, but it's not")
 	}
 }
+
+func TestRecursiveUnmount_SubMountFailsToUnmount(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skip("root required")
+	}
+
+	tmp := t.TempDir()
+
+	dir := "sub1/sub2/sub3"
+	err := os.MkdirAll(path.Join(tmp, dir), 0o700)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	grandchild := path.Join(tmp, dir)
+	child := filepath.Dir(grandchild)
+	parent := filepath.Dir(child)
+
+	// Create a set of mounts that should result in RecursiveUnmount failure,
+	// caused by the fact that the grandchild mount is shadowed by the child mount,
+	// and the child mount is shadowed by the parent mount. So. these two mounts
+	// are listed in mountinfo, but since they are unreachable, unmount will fail.
+	toMount := []string{grandchild, child, parent}
+	for _, dir := range toMount {
+		dir := dir
+		if err := Mount("tmpfs", dir, "tmpfs", ""); err != nil {
+			t.Fatal(err)
+		}
+		//nolint:errcheck
+		defer Unmount(dir)
```

Comment:
Didn't leave it as a comment earlier as it's a minor nit, but if you're updating, perhaps put the `//nolint` after the line; slightly cleaner, and makes it slightly clearer which line the `//nolint` applies to.

```go
defer Unmount(dir) //nolint:errcheck
```